wemake_python_styleguide/visitors/ast/complexity/access.py

```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+
+import ast
+from itertools import takewhile
+from typing import Set, cast
+
+from typing_extensions import final
+
+from wemake_python_styleguide.logic.prop_access import parts
+from wemake_python_styleguide.types import AnyAccess
+from wemake_python_styleguide.violations.complexity import (
+    TooDeepAccessViolation,
+)
+from wemake_python_styleguide.visitors.base import BaseNodeVisitor
+
+
+@final
+class AccessVisitor(BaseNodeVisitor):
+    """Counts access number for expressions."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Keeps visited accesses to not visit them again."""
+        super().__init__(*args, **kwargs)
+        self._visited_accesses: Set[AnyAccess] = set()
+
+    def visit_Subscript(self, node: ast.Attribute) -> None:
+        """
+        Checks subscript access number.
+
+        Raises:
+            TooDeepAccessViolation
+
+        """
+        self._check_consecutive_access_number(node)
+        self.generic_visit(node)
+
+    def visit_Attribute(self, node: ast.Subscript) -> None:
+        """
+        Checks attribute access number.
+
+        Raises:
+            TooDeepAccessViolation
+
+        """
+        self._check_consecutive_access_number(node)
+        self.generic_visit(node)
+
+    def _is_any_access(self, node: ast.AST) -> bool:
+        return isinstance(node, (ast.Attribute, ast.Subscript))
+
+    def _check_consecutive_access_number(self, node: AnyAccess) -> None:
+        if node in self._visited_accesses:
+            return
+
+        consecutive_access = cast(Set[AnyAccess], set(takewhile(
```

Comment:
Why do we need a `cast` here? Is it possible to have: `consecutive_access: Set[AnyAccess] = set(takewhile(..))` ?

Reply:
`mypy` complains on this `consecutive_access: Set[AnyAccess] = set(takewhile(..))`:
```
wemake_python_styleguide/visitors/ast/complexity/access.py:59: error: Argument 2 to "takewhile" has incompatible type "Iterable[AST]"; expected "Iterable[Union[Attribute, Subscript]]"
```

`parts` returns `Iterator[ast.AST]`, so I have to [downcast to subtype](https://mypy.readthedocs.io/en/latest/casts.html) explicitly or use `Set[ast.AST]` instead. But here I should use specific type, not generic `ast.AST`

Reply:
How do you think: can we improve `paths` types?

Reply:
`paths` returns `ast.expr`, because the function depends on child node attributes: `Attribute.value`, `Subscript.value` and `Call.func`, which have `ast.expr` types.

I don't know how to improve it.